tests/unit/model/relationships/has-many-test.js

```diff
@@ -895,6 +895,570 @@ test('it is possible to add a new item to a relationship', function(assert) {
+  });
+
+  const Dog = DS.Model.extend({
+    name: DS.attr('string')
+  });
+
+  let env = setupStore({
+    person: Person,
+    dog: Dog
+  });
+  env.adapter.shouldBackgroundReloadRecord = () => false;
+  env.adapter.deleteRecord = () => {
+    return null;
+  };
+
+  let { store } = env;
+
+  run(() => {
+    store.push({
+      data: {
+        type: 'person',
+        id: '1',
+        attributes: {
+          name: 'Chris Thoburn'
+        },
+        relationships: {
+          dog: {
+            data: { type: 'dog', id: '1' }
+          }
+        }
+      },
+      included: [
+        {
+          type: 'dog',
+          id: '1',
+          attributes: {
+            name: 'Shenanigans'
+          }
+        },
+        {
+          type: 'dog',
+          id: '2',
+          attributes: {
+            name: 'Rambunctious'
+          }
+        }
+      ]
+    });
+  });
+
+  const person = store.peekRecord('person', '1');
+  let dog = run(() => person.get('dog'));
+  const shen = store.peekRecord('dog', '1');
+  const rambo = store.peekRecord('dog', '2');
+
+  assert.ok(dog === shen, 'precond - the belongsTo points to the correct dog');
+  assert.equal(get(dog, 'name'), 'Shenanigans', 'precond - relationships work');
+
+  run(() => {
+    person.set('dog', rambo);
+  });
+
+  dog = person.get('dog');
+  assert.equal(dog, rambo, 'precond2 - relationship was updated');
+
+  return run(() => {
+    return rambo.destroyRecord({})
+      .then(() => {
+        rambo.unloadRecord();
+
+        dog = person.get('dog');
+        assert.equal(dog, null, 'The current state of the belongsTo was clearer');
+      });
+  });
+});
+
+/*
+  This test, when passing, affirms that a known limitation of ember-data still exists.
+
+  When pushing new data into the store, ember-data is currently incapable of knowing whether
+  a relationship has been persisted. In order to update relationship state effectively, ember-data
+  blindly "flushes canonical" state, removing any `currentState` changes. A delete that sideloads
+  the parent record's hasMany is a situation in which this limitation will be encountered should other
+  local changes to the relationship still exist.
+ */
+test('[ASSERTS KNOWN LIMITATION STILL EXISTS] returning new hasMany relationship info from a delete clears local state', function(assert) {
+  assert.expect(4);
+
+  const Person = DS.Model.extend({
+    name: DS.attr('string'),
+    pets: DS.hasMany('pet', { async: false, inverse: null })
+  });
+
+  const Pet = DS.Model.extend({
+    name: DS.attr('string'),
+    person: DS.belongsTo('person', { async: false, inverse: null })
+  });
+
+  let env = setupStore({
+    person: Person,
+    pet: Pet
+  });
+  env.adapter.shouldBackgroundReloadRecord = () => false;
+  let _super = env.serializer.normalize;
+  env.serializer.normalize = function(klass, payload) {
+    // this is required, so that the backported test actually has meaning on the release branch
```

Comment:
this was added to make the back-ported test make sense, without having to refactor a bunch of tests